File: python/22.11/290.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        a = dict()
        b = s.split()
        if len(pattern) != len(b):
            return False
        c = dict()
        for i in range(len(pattern)):
            if a.get(pattern[i]):
                if a.get(pattern[i]) != b[i]:
                    return False
            else:
                if not c.get(b[i]):
                    c[b[i]] = True
                    a[pattern[i]] = b[i]
                else:
                    return False
        return True

File: typescript/23.08/290.ts
function wordPattern(pattern: string, s: string): boolean {
    const [mem, mem2] = [{}, {}]
    let t = s.split(" ")
    if (pattern.length !== t.length) return false

    for (let i=0; i<pattern.length; i++) {
        if (t[i] === "constructor") {
            return true
        }
        if (!mem[pattern[i]]) {
            mem[pattern[i]] = t[i]
        } else {
            if (mem[pattern[i]] !== t[i]) {
                return false
            }
        }
        if (!mem2[t[i]]) {
            mem2[t[i]] = pattern[i]
        } else {
            if (mem2[t[i]] !== pattern[i]) {
                return false
            }
        }
    }

    return true
};
